sycl/doc/design/OptionalDeviceFeatures.md

```diff
@@ -913,211 +913,253 @@ aspects in the array.
+for an Intel GPU device that has not yet been publicly released.  When this
+happens, the configuration file causes the driver to pass
+`clang-offload-wrapper` an option of the form `-target=intel_gpu_XXX` where the
+"XXX" represents the Intel GMDID of the device.
+
+**TODO**: We are considering moving away from the `-fsycl-targets` option and
+using the clang option `--offload-arch` instead.  If we do this, the target
+device names might change too in order to align with the naming style used by
+other offload architectures.  I presume we want to use the same strings for
+both `--offload-arch` and for the `-target` option to `clang-offload-wrapper`.
+Therefore, the target names presented above might change.
+
+**TODO**: The handling of the SPIR-V offload bundle requires some more thought.
+In the current design, there are two different types of SPIR-V modules, which
+are not exactly the same.
+
+If the user compiles with `-fsycl-targets=spir64` (which is the default), the
+device compiler is invoked with `-triple spir64-unknown-unknown`, and the
+generated IR is converted to SPIR-V.  This type of SPIR-V module is wrapped
+into the host executable via `clang-offload-wrapper -target=spir64` (i.e. the
+offload bundle has type "spir64").
+
+However, if the user AOT compiles for an Intel GPU with
+`-fsycl-targets=spir64_gen`, the device compiler is invoked with
+`-triple spir64_gen-unknown-unknown`, which is a different triple from the case
+above.  This IR is also converted to SPIR-V and then sent to the `ocloc` tool.
+The `ocloc` tool compiles the SPIR-V to native code but also embeds the SPIR-V
+into the ocloc output.  This SPIR-V is used by the device driver if the
+application is run on a GPU device that does not match any of the AOT targets.
+
+In the new design, `ocloc` does not embed SPIR-V in its output because we
+invoke it with `-exclude_ir`.  Instead, DPC++ is responsible for embedding the
+SPIR-V in the host executable, but should that SPIR-V be generated with the
+triple `spir64-unknown-unknown` or the triple `spir64_gen-unknown-unknown`?
+We could generate both, but that seems wasteful.  If we generate the SPIR-V
+with `spir64-unknown-unknown`, does that mean we need to run the device
+compiler twice, once with `spir64_gen-unknown-unknown` (to generate SPIR-V that
+is fed to `ocloc`) and once with `spir64-unknown-unknown` (to generate SPIR-V
+that is embedded in the host executable)?  This also seems wasteful because
+it requires running the device compiler twice whenever the user AOT compiles
+for an Intel GPU.
+
+An ideal solution would be to always generate SPIR-V using the triple
+`spir64-unknown-unknown`, and use this SPIR-V as input to both `ocloc` and
+`opencl-aot`.  This would allow us to run the device compiler just once even if
+the user AOT compiles for both Intel GPU and CPU.  However, this means that the
+SPIR-V is generated using a different triple than we use today.  It's unclear
+what ramification this has.
```

Comment:
Tagging some front-end people here for their perspective: @tahonermann @elizabethandrews @premanandrao 

What are the front-end ramifications if we compile to SPIR-V using the triple `spir64-unknown-unknown` vs `spir64_gen-unknown-unknown` vs. `spir64_x86_64-unknown-unknown`?

Reply:
@gmlueck, I can't think of any FE impact in particular.  Did you have anything specific in mind?  If there are any changes in size/layout of types, this would be more of a downstream component impact than in FE itself.   Maybe I am missing the point of your question.

Reply:
I might be wrong, but I had the impression that the triple had some effect on the front-end, like enabling different builtins.  If that's not the case, it would be good news.

Reply:
I can't comment specifically on `spir64` vs `spir64_gen` vs `spir64_x86_64`, but in general, such differences certainly can effect the FE in terms of predefined macros, enabled features, builtins, etc... Some investigation would be needed to discover what specific differences would be enabled in these cases.

Reply:
I think it would be worthwhile to do this investigation.  If there is a difference, I don't think we are documenting it well to our users.  Most people assume that the only difference between JIT compilation (`spir64`) vs. AOT compilation (`spir64_gen` or `spir64_x86_64`) is a tradeoff between portability and runtime performance -- JIT mode is more portable to different h/w but has a higher runtime cost due to the JIT compilation step.

However, if the FE behaves differently in these three modes, then that means there is also a functional difference between JIT compilation and AOT compilation.

Reply:
The idea of having a single device triple for all targets would be ideal, and also simplifies how things are represented in general, allowing for specific device targets to be specified via an `--offload-arch` (community reference here) value without depending on the triple arch_subarch.